File: scripts/extract-scenes-51-100.py
#!/usr/bin/env python3
import json
import re
from pathlib import Path

# Read the outline file
outline_path = Path(__file__).parent.parent / 'data' / 'l_outline.json'
with open(outline_path, 'r') as f:
    outline_data = json.load(f)

# Extract all scenes with their chapter context
all_scenes = []

def extract_scenes_from_book(book_data, trilogy_name, book_name):
    """Extract scenes from a book's task_masters structure"""
    task_masters = book_data.get('task_masters', {})

    for tm_key, tm_value in task_masters.items():
        major_task_groups = tm_value.get('major_task_groups', {})

        for mtg_key, mtg_value in major_task_groups.items():
            specific_task_groups = mtg_value.get('Specific_task_groups', {})

            for stg_key, chapter in specific_task_groups.items():
                chapter_id = chapter.get('id', '')
                scenes = chapter.get('scenes', [])

                for scene in scenes:
                    scene_with_context = {
                        'scene_index': 0,  # Will be set after sorting
                        'trilogy': trilogy_name,
                        'book': book_name,
                        'chapter_id': chapter_id,
                        'scene_number': scene.get('scene_number', 0),
                        'title': scene.get('title', ''),
                        'pov': scene.get('pov', ''),
                        'setup': scene.get('setup', ''),
                        'description': scene.get('description', ''),
                        'focus': scene.get('focus', ''),
                        'current_internal_conflict': scene.get('internal_conflict') or scene.get('internalConflict', ''),
                        'characters_involved': scene.get('characters', []),
                        'scene_data': scene
                    }
                    all_scenes.append(scene_with_context)

# Navigate through the structure
series = outline_data.get('SelfImprovementSeries', {})
books_section = series.get('Books', {})
trilogies = books_section.get('trilogies', {})

# Process each trilogy
for trilogy_key, trilogy_data in trilogies.items():
    trilogy_books = trilogy_data.get('trilogy_books', {})

    # Process each book in the trilogy
    for book_key, book_data in trilogy_books.items():
        extract_scenes_from_book(book_data, trilogy_key, book_key)

# Sort scenes by chapter ID (EA-XXX) and then by scene number
def get_chapter_num(chapter_id):
    match = re.search(r'EA-(\d+)', chapter_id)
    return int(match.group(1)) if match else 0

all_scenes.sort(key=lambda s: (get_chapter_num(s['chapter_id']), s['scene_number']))

# Assign scene indices
for index, scene in enumerate(all_scenes):
    scene['scene_index'] = index + 1

# Extract scenes 51-100
scenes_51_to_100 = all_scenes[50:100]  # 0-indexed, so 50-99

print(f"Total scenes in outline: {len(all_scenes)}")
print(f"Extracting scenes 51-100 ({len(scenes_51_to_100)} scenes)")

# Save the extracted scenes for analysis
output_path = Path(__file__).parent / 'scenes-51-100-raw.json'
with open(output_path, 'w') as f:
    json.dump(scenes_51_to_100, f, indent=2)

print(f"Saved raw scenes to: {output_path}")

if scenes_51_to_100:
    print(f"\nFirst scene:")
    first = scenes_51_to_100[0]
    print(f"  Scene {first['scene_index']}: {first['chapter_id']} - {first['title']}")

    print(f"\nLast scene:")
    last = scenes_51_to_100[-1]
    print(f"  Scene {last['scene_index']}: {last['chapter_id']} - {last['title']}")

File: scripts/extract-scenes-51-100.ts
import * as fs from 'fs';
import * as path from 'path';

interface Scene {
  scene_number: number;
  title?: string;
  setup?: string;
  description?: string;
  focus?: string;
  pov?: string;
  internal_conflict?: string;
  internalConflict?: string;
  characters?: string[];
  [key: string]: any;
}

interface Chapter {
  chapter_id: string;
  scenes: Scene[];
  [key: string]: any;
}

interface Outline {
  chapters: Chapter[];
  [key: string]: any;
}

// Read the outline file
const outlinePath = path.join(__dirname, '..', 'data', 'l_outline.json');
const outlineData: Outline = JSON.parse(fs.readFileSync(outlinePath, 'utf-8'));

// Extract all scenes with their chapter context
interface SceneWithContext {
  scene_index: number;
  chapter_id: string;
  scene_number: number;
  title: string;
  pov: string;
  setup: string;
  description: string;
  focus: string;
  current_internal_conflict: string;
  characters_involved: string[];
  scene_data: Scene;
}

const allScenes: SceneWithContext[] = [];

// Collect all scenes from all chapters
outlineData.chapters.forEach((chapter: Chapter) => {
  if (chapter.scenes && Array.isArray(chapter.scenes)) {
    chapter.scenes.forEach((scene: Scene) => {
      const sceneWithContext: SceneWithContext = {
        scene_index: 0, // Will be set after sorting
        chapter_id: chapter.chapter_id || '',
        scene_number: scene.scene_number || 0,
        title: scene.title || '',
        pov: scene.pov || '',
        setup: scene.setup || '',
        description: scene.description || '',
        focus: scene.focus || '',
        current_internal_conflict: scene.internal_conflict || scene.internalConflict || '',
        characters_involved: scene.characters || [],
        scene_data: scene
      };
      allScenes.push(sceneWithContext);
    });
  }
});

// Sort scenes by chapter ID (EA-XXX) and then by scene number
allScenes.sort((a, b) => {
  // Extract chapter number from EA-XXX format
  const getChapterNum = (id: string) => {
    const match = id.match(/EA-(\d+)/);
    return match ? parseInt(match[1]) : 0;
  };

  const chapterA = getChapterNum(a.chapter_id);
  const chapterB = getChapterNum(b.chapter_id);

  if (chapterA !== chapterB) {
    return chapterA - chapterB;
  }

  return a.scene_number - b.scene_number;
});

// Assign scene indices
allScenes.forEach((scene, index) => {
  scene.scene_index = index + 1;
});

// Extract scenes 51-100
const scenes51to100 = allScenes.slice(50, 100); // 0-indexed, so 50-99

console.log(`Total scenes in outline: ${allScenes.length}`);
console.log(`Extracting scenes 51-100 (${scenes51to100.length} scenes)`);

// Save the extracted scenes for analysis
const outputPath = path.join(__dirname, 'scenes-51-100-raw.json');
fs.writeFileSync(outputPath, JSON.stringify(scenes51to100, null, 2));

console.log(`Saved raw scenes to: ${outputPath}`);
console.log('\nFirst scene:');
console.log(`  Scene ${scenes51to100[0]?.scene_index}: ${scenes51to100[0]?.chapter_id} - ${scenes51to100[0]?.title}`);
console.log('\nLast scene:');
console.log(`  Scene ${scenes51to100[scenes51to100.length - 1]?.scene_index}: ${scenes51to100[scenes51to100.length - 1]?.chapter_id} - ${scenes51to100[scenes51to100.length - 1]?.title}`);
